### PR TITLE
Explicitly pass QgsProject object when dealing with expression contexts

### PR DESCRIPTION
**Pull request by wonder-sk**
Continued effort to reduce number of uses of QgsProject as singleton...

### CODE REVIEW

File: src/app/qgsrulebasedlabelingwidget.cpp [hunk 1]
```diff
@@ -27,6 +28,27 @@
...
+  if ( mapCanvas )
+  {
+    scopes << QgsExpressionContextUtils::mapSettingsScope( mapCanvas->mapSettings() )
+    << new QgsExpressionContextScope( mapCanvas->expressionContextScope() );
+  }
+  else
+  {
+    scopes << QgsExpressionContextUtils::mapSettingsScope( QgsMapSettings() );
+  }
+  scopes << QgsExpressionContextUtils::layerScope( layer );
```
Comment (nyalldawson) [inline src/app/qgsrulebasedlabelingwidget.cpp]: Should we check if layer & project are null here?

**Review by nyalldawson — COMMENTED** [review]
I'm fairly certain processing uses these methods too and will probably need updating to match. But otherwise looks good to me!

File: src/app/qgsrulebasedlabelingwidget.cpp (hunk 1 above)
Comment (wonder-sk) [inline src/app/qgsrulebasedlabelingwidget.cpp]: I was thinking the same - I have just copied the code from the original places without any changes to logic. There are some other places in the code that check whether layer is not null and add layer scope only we have a valid layer. I am happy to add not-null check for either project and layer as that makes more sense than just adding dummy scopes...

**Comment by wonder-sk** [comment]
Thanks @nyalldawson for having a look - I will check processing code... (this is where I really like c++: much easier to check class/method usages!)

**Pull request merged**